keps/sig-node/4680-add-resource-health-to-pod-status/README.md

```diff
@@ -0,0 +1,451 @@
+
+The proposal is to keep this structure as-is to simplify parsing of well-known ResourceList data type by various consumers. Typical scenario would be to compare if the `AllocatedResources` match the desired state.
+
+The proposal is to introduce an additional field:
+
+```
+// AllocatedResourcesStatus represents the status of various resources
+// allocated for this Pod.
+AllocatedResourcesStatus ResourcesStatus 
+
+type ResourcesStatus map[ResourceName]ResourceStatus
+
+type ResourceStatus struct {
+    // map of unique Resource ID to its health.
+    // At a minimum, ResourceID must uniquely identify the Resource
+    // allocated to the Pod on the Node for the lifetime of a Pod.
+    // See ResourceID type for it's definition.
+    Resources map[ResourceID] ResourceHealth
+    
+    // allow to extend this struct in future with the overall health fields or things like Device Plugin version
+}
+
+// ResourceID is calculated based on the source of this resource health information.
+// For DevicePlugin:
+//   deviceplugin:DeviceID, where DeviceID is from the Device structure of DevicePlugin's ListAndWatchResponse type: https://github.com/kubernetes/kubernetes/blob/eda1c780543a27c078450e2f17d674471e00f494/staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1alpha/api.proto#L61-L73
+// DevicePlugin ID is usually a constant for the lifetime of a Node and typically can be used to uniquely identify the device on the node.
+// For DRA:
+//   dra:<driver name>[/<pool name>]/<device name>: such a device can be looked up in the information published by that DRA driver to learn more about it. It is designed to be globally unique in a cluster.
+type ResourceID string
+
+type ResourceHealth struct {
+    // List of conditions with the transition times
+    Conditions []ResourceHealthCondition
+}
+
+// This condition type is replicating other condition types exposed by various status APIs
+type ResourceHealthCondition struct {
+    // can be one of:
+    //  - Healthy: operates as normal
+    //  - Unhealthy: reported unhealthy. We consider this a temporary health issue
+    //               since we do not mechanism today to distinguish
+    //               temporary and permanent issues.
+    //  - Unknown: The status cannot be determined.
+    //             For example, Device Plugin got unregistered and hasn't been re-registered since.
+    //
+    // In future we may want to introduce the PermanentlyUnhealthy Status.
+    Type string 
+
+    // Status of the condition, one of True, False, Unknown.
+    Status ConditionStatus
+    // The last time the condition transitioned from one status to another.
+    // +optional
+    LastTransitionTime metav1.Time
+    // The reason for the condition's last transition.
+    // +optional
+    Reason string
+    // A human readable message indicating details about the transition.
+    // +optional
+    Message string
```

Comment:
shadow prod readiness review

optionals should be a pointer type [ref](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required)?

Reply:
I would think we can address this as part of implementation. As I mentioned before, this is copied. So I want to keep it consistent with other types